pkg/segment/query/querystatus.go

```diff
@@ -278,19 +279,32 @@ func canRunQuery() bool {
 	return activeQueries < MAX_RUNNING_QUERIES
 }
 
+func initiateRunQuery(segsRLockFunc, segsRUnlockFunc func()) {
+	waitingQueriesLock.Lock()
+	if len(waitingQueries) == 0 {
+		waitingQueriesLock.Unlock()
+		time.Sleep(PULL_QUERY_INTERVAL)
+		return
```

Comment:
nit: sleeping before a return seems really odd to me. I see you got this from refactoring, where the sleep was before a `continue`. Here's what I would suggest:
1. Make this function return `*wsData` instead of nothing, and remove the Sleep() here
2. If the caller of this function gets nil, then sleep and continue
3. Call `runQuery()` in the caller, so that in this function you can make the locking a bit easier by deferring the unlock to avoid multiple unlock paths